src/librbd/internal.cc

```diff
@@ -689,6 +689,64 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
     return 0;
   }
 
+  int list_children2(ImageCtx *ictx,
+                     vector<children_info_t>& names) {
+    CephContext *cct = ictx->cct;
+    ldout(cct, 20) << "children list " << ictx->name << dendl;
+
+    int r = ictx->state->refresh_if_required();
+    if (r < 0) {
+      return r;
+    }
+
+    RWLock::RLocker l(ictx->snap_lock);
+    ParentSpec parent_spec(ictx->md_ctx.get_id(), ictx->id, ictx->snap_id);
+    map< pair<int64_t, string>, set<string> > image_info;
+
+    r = api::Image<>::list_children(ictx, parent_spec, &image_info);
+    if (r < 0) {
+      return r;
+    }
+
+    Rados rados(ictx->md_ctx);
+    for ( auto &info : image_info){
+      IoCtx ioctx;
+      r = rados.ioctx_create2(info.first.first, ioctx);
+      if (r < 0) {
+        lderr(cct) << "Error accessing child image pool " << info.first.second
+                   << dendl;
+        return r;
+      }
+
+      children_info_t children;
+      children.pool_name = info.first.second;
+
+      for (auto &id_it : info.second) {
+	string name;
+	r = cls_client::dir_get_name(&ioctx, RBD_DIRECTORY, id_it, &name);
+	if (r < 0 && r != -ENOENT) {
+          lderr(cct) << "Error looking up name for image id " << id_it
+                     << " in pool " << info.first.second << dendl;
+          return r;
+        } else if (r == -ENOENT) {
```

Comment:
Nit: execute the ```-ENOENT``` handling first. If it fails w/ -EOENT, you should allow it to fall through to the generic handling so that the error message is "error looking up name for image id XYZ in pool" instead of "in rbd trash".